src/3d/mesh/qgsmesh3dgeometry_p.cpp

```diff
@@ -0,0 +1,421 @@
+      return false;
+    }
+
+    QT3D_FUNCTOR( MeshPlaneVertexBufferFunctor )
+
+  private:
+    QgsTriangularMesh mMesh;
+    QVector<double> mVerticaleMagnitude;
+    QVector<double> mScalarMagnitude;
+    QgsRectangle mExtent;
+    float mVertScale;
+
+};
+
+QgsMesh3dGeometry_p::QgsMesh3dGeometry_p(
+  const QgsTriangularMesh &mesh,
+  const QgsRectangle &extent,
+  float verticaleScale,
+  QgsMesh3dGeometry_p::QNode *parent ):
+  QGeometry( parent ),
+  mTriangularMesh( mesh ),
+  mExtent( extent ),
+  mVertScale( verticaleScale )
+{
+  init();
+}
+
+void QgsMesh3dGeometry_p::init()
+{
+  mPositionAttribute = new QAttribute( this );
+  mNormalAttribute = new QAttribute( this );
+  mTexCoordAttribute = new QAttribute( this );
+  mIndexAttribute = new QAttribute( this );
+  mVertexBuffer = new Qt3DRender::QBuffer( Qt3DRender::QBuffer::VertexBuffer, this );
+  mIndexBuffer = new Qt3DRender::QBuffer( Qt3DRender::QBuffer::IndexBuffer, this );
+
+
+  const int stride = ( 3 /*position*/ +
+                       2 /*texture*/  +
+                       3 /*normale*/ ) * sizeof( float );
+
+  const uint nVerts = uint( mTriangularMesh.vertices().count() );
+
+  mPositionAttribute->setName( QAttribute::defaultPositionAttributeName() );
+  mPositionAttribute->setVertexBaseType( QAttribute::Float );
+  mPositionAttribute->setVertexSize( 3 );
+  mPositionAttribute->setAttributeType( QAttribute::VertexAttribute );
+  mPositionAttribute->setBuffer( mVertexBuffer );
+  mPositionAttribute->setByteStride( stride );
+  mPositionAttribute->setCount( nVerts );
+
+  mTexCoordAttribute->setName( QAttribute::defaultTextureCoordinateAttributeName() );
+  mTexCoordAttribute->setVertexBaseType( QAttribute::Float );
+  mTexCoordAttribute->setVertexSize( 2 );
+  mTexCoordAttribute->setAttributeType( QAttribute::VertexAttribute );
+  mTexCoordAttribute->setBuffer( mVertexBuffer );
+  mTexCoordAttribute->setByteStride( stride );
+  mTexCoordAttribute->setByteOffset( 3 * sizeof( float ) );
+  mTexCoordAttribute->setCount( nVerts );
+
+  mNormalAttribute->setName( QAttribute::defaultNormalAttributeName() );
+  mNormalAttribute->setVertexBaseType( QAttribute::Float );
+  mNormalAttribute->setVertexSize( 3 );
+  mNormalAttribute->setAttributeType( QAttribute::VertexAttribute );
+  mNormalAttribute->setBuffer( mVertexBuffer );
+  mNormalAttribute->setByteStride( stride );
+  mNormalAttribute->setByteOffset( 5 * sizeof( float ) );
+  mNormalAttribute->setCount( nVerts );
+
+  mIndexAttribute->setAttributeType( QAttribute::IndexAttribute );
+  mIndexAttribute->setVertexBaseType( QAttribute::UnsignedInt );
+  mIndexAttribute->setBuffer( mIndexBuffer );
+
+  // Each primitive has 3 vertives
+  mIndexAttribute->setCount( uint( mTriangularMesh.triangles().count() ) * 3 );
+
+  mVertexBuffer->setData( MeshPlaneVertexBufferFunctor( mTriangularMesh, mExtent, mVertScale )() );
+  mIndexBuffer->setData( MeshPlaneIndexBufferFunctor( mTriangularMesh )() );
```

Comment:
I would suggest to remove the functors and simply call the create*() functions directly here - to simplify the code

Reply:
This the way used in the example of Qt3d, and the used for `DemTerrainTileGeometry` ... but you are right, it would simplify the code. Or maybe, it is better to use setDataGenerator(), so the work could done asynchronously (https://doc.qt.io/qt-5/qt3drender-qbuffer.html#details)

Reply:
I am quite easy with this - feel free to also skip this point...